lib/HooksHandler.php

```diff
@@ -21,118 +21,242 @@
-				$params['message'] = (string) $l->t('Expiration date is required');
+				$params['message'] = (string) $this->l10n->t('Expiration date is required');
 			}
 
 			// $date is the max expiration date
 			if ($date < $params['expirationDate']) {
 				$params['accepted'] = false;
-				$params['message'] = (string) $l->n('Expiration date cannot exceed %n day', 'Expiration date cannot exceed %n days', $days);
+				$params['message'] = (string) $this->l10n->n('Expiration date cannot exceed %n day', 'Expiration date cannot exceed %n days', $days);
 			}
 		}
 	}
 
+	public function saveOldPassword(GenericEvent $event) {
+		$this->fixDI();
+
+		$user = $this->getUser($event);
+		$password = $event->getArgument('password');
+
+		$oldPassword = new OldPassword();
+		$oldPassword->setUid($user->getUID());
+		$oldPassword->setPassword($this->hasher->hash($password));
+		$oldPassword->setChangeTime($this->timeFactory->getTime());
+		$this->oldPasswordMapper->insert($oldPassword);
+	}
+
 	/**
-	 * @return array
+	 * Flags the session to require a password change
 	 */
-	private static function loadConfiguration() {
-		$appValues = [
-			'spv_min_chars_checked' => false,
-			'spv_min_chars_value' => 8,
-			'spv_lowercase_checked' => false,
-			'spv_lowercase_value' => 1,
-			'spv_uppercase_checked' => false,
-			'spv_uppercase_value' => 1,
-			'spv_numbers_checked' => false,
-			'spv_numbers_value' => 1,
-			'spv_special_chars_checked' => false,
-			'spv_special_chars_value' => 1,
-			'spv_def_special_chars_checked' => false,
-			'spv_def_special_chars_value' => '#!',
-			'spv_password_history_checked' => false,
-			'spv_password_history_value' => 3,
-			'spv_expiration_password_checked' => false,
-			'spv_expiration_password_value' => 7,
-			'spv_expiration_nopassword_checked' => false,
-			'spv_expiration_nopassword_value' => 7,
-		];
-
-		$configValues = [];
-		foreach ($appValues as $key => $default) {
-			$configValues[$key] = \OC::$server->getConfig()->getAppValue('password_policy', $key, $default);
-		}
-		return $configValues;
+	protected function forcePasswordChange() {
+		$this->session->set('password_policy.forcePasswordChange', true);
 	}
 
 	/**
-	 * @param IUser  $user
-	 * @param string $password
+	 * check we really get an object, die hard in case we don't
+	 *
+	 * @param GenericEvent $event
+	 * @return IUser
 	 */
-	public static function saveOldPassword($user, $password) {
-		$dbMapper = new OldPasswordMapper(\OC::$server->getDatabaseConnection());
-		$oldPassword = new OldPassword();
-		$oldPassword->setUid($user->getUID());
-		$oldPassword->setPassword(\OC::$server->getHasher()->hash($password));
-		$oldPassword->setChangeTime(\OC::$server->getTimeFactory()->getTime());
-		$dbMapper->insert($oldPassword);
+	private function getUser(GenericEvent $event) {
+		$user = $event->getArgument('user');
+		if(!$user instanceof IUser) {
+			throw new \UnexpectedValueException("user '$user' not an instance of IUser");
```

Comment:
"_The user, '$user', is not an instance of IUser._"

Reply:
Technical exception. should never happen, but we check it to detect broken code flows. @settermjd why the two commes? You did not add them in the other text changes.